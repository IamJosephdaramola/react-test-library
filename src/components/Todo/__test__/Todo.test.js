import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Todo from "../Todo";

const MockedTodo = () => (
  <BrowserRouter>
    <Todo />
  </BrowserRouter>
);

const addTasks = (tasks) => {
  const inputElement = screen.getByPlaceholderText(/Add a new task here.../i);
  const buttonElement = screen.getByRole("button", { name: /Add/i });

  tasks.forEach((task) => {
    fireEvent.change(inputElement, {
      target: { value: task },
    });
    fireEvent.click(buttonElement);
  });
};

describe("Todo", () => {
  afterEach(cleanup);

  test("should take a snapshot of the Todo", () => {
    const { asFragment } = render(<MockedTodo />);

    expect(asFragment(<MockedTodo />)).toMatchSnapshot();
  });

  test("should render the same text passed into title prop", () => {
    render(<MockedTodo />);

    addTasks(["Relocate to the UK"]);
    const outputElement = screen.getByText(/Relocate to the UK/i);

    expect(outputElement).toBeInTheDocument();
  });

  test("should render multiple items in the todo list", () => {
    render(<MockedTodo />);

    const tasks = ["Relocate to UK", "Wash plates", "clothes", "makes"];

    addTasks(tasks);
    const outputElement = screen.getAllByTestId("todo");

    expect(outputElement.length).toBe(tasks.length);
  });

  test("task should not have the completed task when initially rendered", () => {
    render(<MockedTodo />);

    addTasks(["Go shopping"]);

    const divElement = screen.getByTestId("todo");

    expect(divElement).not.toHaveClass("todo-item-active");
  });

  test("task should  have the completed task when clicked", () => {
    render(<MockedTodo />);

    addTasks(["Go shopping"]);

    const divElement = screen.getByTestId("todo");

    fireEvent.click(divElement);

    expect(divElement).toHaveClass("todo-item-active");
  });
});
