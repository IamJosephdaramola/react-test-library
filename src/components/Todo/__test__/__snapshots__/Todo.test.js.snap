// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should take a snapshot of the Todo 1`] = `
<DocumentFragment>
  <div
    class="todo"
  >
    <h1
      class="header"
      data-testid="header-1"
    >
      Todo
    </h1>
    <div
      class="input-container"
    >
      <input
        class="input"
        placeholder="Add a new task here..."
        value=""
      />
      <button
        class="add-btn"
      >
        Add
      </button>
    </div>
    <div
      class="todolist-container"
    >
      <div
        class="todos-container"
      >
        <div />
      </div>
      <div>
        <div
          class="todo-footer"
        >
          <p>
            0 tasks left
          </p>
          <a
            href="/followers"
          >
            Followers
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
